drivers/bus/isapnp/pdo.c

```diff
@@ -76,82 +82,412 @@ IsaPdoQueryCapabilities(
+    _Inout_ PIRP Irp)
 {
-    Irp->IoStatus.Information |= PNP_DEVICE_NOT_DISABLEABLE;
-    return STATUS_SUCCESS;
+    PAGED_CODE();
+
+    if (PdoExt->Flags & ISAPNP_READ_PORT_NEED_REBALANCE)
+    {
+        Irp->IoStatus.Information |= PNP_DEVICE_NOT_DISABLEABLE |
+                                     PNP_DEVICE_RESOURCE_REQUIREMENTS_CHANGED |
+                                     PNP_DEVICE_FAILED;
+        return STATUS_SUCCESS;
+    }
+    else if (PdoExt->SpecialFiles > 0)
+    {
+        Irp->IoStatus.Information |= PNP_DEVICE_NOT_DISABLEABLE;
+        return STATUS_SUCCESS;
+    }
+
+    return Irp->IoStatus.Status;
 }
 
+static
+CODE_SEG("PAGE")
 NTSTATUS
-NTAPI
 IsaPdoQueryId(
-    IN PISAPNP_PDO_EXTENSION PdoExt,
-    IN PIRP Irp,
-    IN PIO_STACK_LOCATION IrpSp)
+    _In_ PISAPNP_PDO_EXTENSION PdoExt,
+    _Inout_ PIRP Irp,
+    _In_ PIO_STACK_LOCATION IrpSp)
+{
+    PISAPNP_LOGICAL_DEVICE LogDev = PdoExt->IsaPnpDevice;
+    NTSTATUS Status;
+    PWCHAR Buffer, End;
+    size_t CharCount, Remaining;
+
+    PAGED_CODE();
+
+    switch (IrpSp->Parameters.QueryId.IdType)
+    {
+        case BusQueryDeviceID:
+        {
+            CharCount = strlen("ISAPNP\\") + 3 + 4 + sizeof(ANSI_NULL);
+
+            if (LogDev->Flags & ISAPNP_HAS_MULTIPLE_LOGDEVS)
+                CharCount += strlen("_DEV0000");
+
+            Buffer = ExAllocatePoolWithTag(PagedPool,
+                                           CharCount * sizeof(WCHAR),
+                                           TAG_ISAPNP);
+            if (!Buffer)
+                return STATUS_INSUFFICIENT_RESOURCES;
+
+            Status = RtlStringCchPrintfExW(Buffer,
+                                           CharCount,
+                                           &End,
+                                           &Remaining,
+                                           0,
+                                           L"ISAPNP\\%.3S%04x",
+                                           LogDev->VendorId,
+                                           LogDev->ProdId);
+            if (!NT_SUCCESS(Status))
+                goto Cleanup;
+
+            if (LogDev->Flags & ISAPNP_HAS_MULTIPLE_LOGDEVS)
+            {
+                Status = RtlStringCchPrintfExW(End,
+                                               Remaining,
+                                               NULL,
+                                               NULL,
+                                               0,
+                                               L"_DEV%04X",
+                                               LogDev->LDN);
+                if (!NT_SUCCESS(Status))
+                    goto Cleanup;
+            }
+
+            DPRINT("Device ID: '%S'\n", Buffer);
+            break;
+        }
+
+        case BusQueryHardwareIDs:
+        {
+            PWCHAR IdStart;
+
+            CharCount = strlen("ISAPNP\\") + 3 + 4 + sizeof(ANSI_NULL) +
+                        strlen("*") + 3 + 4 + 2 * sizeof(ANSI_NULL);
```

Comment:
Maybe a comment telling what "3" and "4" are for? And that this is going to be a multi-string.